dbt/adapters/default.py

```diff
@@ -0,0 +1,527 @@
+                             target_column.data_type,
+                             new_type,
+                             to_schema,
+                             to_table)
+
+                cls.alter_column_type(profile, to_schema, to_table,
+                                      column_name, new_type, model_name)
+
+    ###
+    # SANE ANSI SQL DEFAULTS
+    ###
+    @classmethod
+    def get_create_schema_sql(cls, schema):
+        return ('create schema if not exists "{schema}"'
+                .format(schema=schema))
+
+    @classmethod
+    def get_create_table_sql(cls, schema, table, columns, sort, dist):
+        fields = ['"{field}" {data_type}'.format(
+            field=column.name, data_type=column.data_type
+        ) for column in columns]
+        fields_csv = ",\n  ".join(fields)
+        dist = cls.dist_qualifier(dist)
+        sort = cls.sort_qualifier('compound', sort)
+        sql = """
+        create table if not exists "{schema}"."{table}" (
+        {fields}
+        )
+        {dist} {sort}
+        """.format(
+            schema=schema,
+            table=table,
+            fields=fields_csv,
+            sort=sort,
+            dist=dist).strip()
+        return sql
+
+    ###
+    # ODBC FUNCTIONS -- these should not need to change for every adapter,
+    #                   although some adapters may override them
+    ###
+    @classmethod
+    def get_default_schema(cls, profile):
+        return profile.get('schema')
+
+    @classmethod
+    def get_connection(cls, profile, name=None, recache_if_missing=True):
+        global connections_in_use
+
+        if name is None:
+            # if a name isn't specified, we'll re-use a single handle
+            # named 'master'
+            name = 'master'
+
+        if connections_in_use.get(name):
+            return connections_in_use.get(name)
+
+        if recache_if_missing is False:
+            raise dbt.exceptions.InternalException(
+                'Tried to get a connection "{}" which does not exist '
+                '(recache_if_missing is off).'.format(name))
+
+        logger.debug('Acquiring new {} connection "{}".'
+                     .format(cls.type(), name))
+
+        connection = cls.acquire_connection(profile, name)
+        connections_in_use[name] = connection
+
+        return cls.get_connection(profile, name)
+
+    @classmethod
+    def total_connections_allocated(cls):
+        global connections_in_use, connections_available
+
+        return len(connections_in_use) + len(connections_available)
+
+    @classmethod
+    def acquire_connection(cls, profile, name):
+        global connections_available, lock
+
+        # we add a magic number, 2 because there are overhead connections,
+        # one for pre- and post-run hooks and other misc operations that occur
+        # before the run starts, and one for integration tests.
+        max_connections = profile.get('threads', 1) + 2
```

Comment:
this connection pool doesn't do any sophisticated retry logic if a new connection is available. it just has a fixed number of connections. if you try to acquire a connection but they are all already in use, you get an exception.

in addition to one thread per model, we need some overheard connections. one is 'master' which is used for pre- and post-run-hooks, getting the list of existing tables before model runs, creating the schema, etc.

the other is for testing, which is kind of dumb.

this code doesn't exactly cap the number of connections to the number of threads, but it does make sure that connections don't grow in an unbounded fashion.